File: meeting_pipeline/pipeline.py
"""Orchestrate one recording: convert -> transcribe -> clean -> summarise -> note."""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from types import SimpleNamespace
from typing import Any

from . import transcribe as transcribe_mod
from . import summarise as summarise_mod
from .cleaning import clean_transcript, segments_from_result
from .config import resolve_path
from .state import State, base_for, wait_until_stable
from .validate import validate_summary

SUPPORTED_EXTENSIONS = {
    ".wav", ".m4a", ".mp3", ".opus", ".ogg", ".flac", ".aac", ".m4b",
    ".mov", ".mp4", ".m4v", ".3gp", ".webm", ".mkv",
}


@dataclass
class ProcessResult:
    status: str
    base: str
    message: str
    note_path: Path | None = None
    transcript_path: Path | None = None


def _default_deps() -> SimpleNamespace:
    return SimpleNamespace(
        convert_to_wav=transcribe_mod.convert_to_wav,
        transcribe=transcribe_mod.transcribe,
        ollama_reachable=summarise_mod.ollama_reachable,
        summarise=summarise_mod.summarise,
    )


def _choose_ollama(cfg: dict, deps: SimpleNamespace) -> tuple[dict, bool]:
    """Return (chosen target dict, deferred). target has url+model."""
    scfg = cfg["summarise"]
    server = scfg["server"]
    local = scfg["local"]
    if scfg.get("backend") == "local":
        return local, False
    if deps.ollama_reachable(server["url"]):
        return server, False
    if scfg.get("allow_local_fallback"):
        return local, False
    return {}, True


def process_one(
    path: Path,
    cfg: dict,
    *,
    deps: Any = None,
    stable_checks: int = 3,
    stable_delay: float = 2.0,
) -> ProcessResult:
    deps = deps or _default_deps()
    recordings_dir = resolve_path(cfg["recordings_dir"])
    base = base_for(recordings_dir, path)

    notes_dir = resolve_path(cfg["notes_dir"])
    work_dir = resolve_path(cfg["work_dir"])
    state = State(work_dir / ".state", notes_dir)

    if state.is_done(base):
        return ProcessResult("skipped", base, "already processed",
                             state.note_path(base))

    if not wait_until_stable(path, checks=stable_checks, delay=stable_delay):
        return ProcessResult("skipped", base, "file still changing")

    if state.is_processing(base):
        return ProcessResult("skipped", base, "already being processed")

    target, deferred = _choose_ollama(cfg, deps)
    if deferred:
        return ProcessResult(
            "deferred_need_local", base,
            "Server Ollama offline; local fallback not allowed yet")

    state.mark_processing(base)
    notes_dir.mkdir(parents=True, exist_ok=True)
    note_path = state.note_path(base)

    try:
        wav_path = work_dir / f"{base}.wav"
        deps.convert_to_wav(path, wav_path)

        result = deps.transcribe(wav_path, cfg["transcribe"])
        segments = segments_from_result(result)
        clean = clean_transcript(segments)
        if not clean.strip():
            state.mark_failed(base, "empty transcript")
            return ProcessResult("failed", base, "empty transcript")

        transcript_path = work_dir / f"{base}.transcript.clean.txt"
        transcript_path.write_text(clean + "\n", encoding="utf-8")

        summary, _raw = deps.summarise(
            clean,
            url=target["url"], model=target["model"],
            options=cfg["summarise"]["options"],
            note_owner=cfg["note_owner"], user_speaker=cfg["user_speaker"],
        )
        note_path.write_text(summary, encoding="utf-8")

        failures = validate_summary(summary)
        if failures:
            state.mark_failed(base, "; ".join(failures))
            return ProcessResult("failed", base, "; ".join(failures), note_path,
                                 transcript_path)

        state.mark_done(base)
        return ProcessResult("done", base, "note written", note_path, transcript_path)

    except Exception as exc:  # transcribe/summarise/convert errors
        state.mark_failed(base, f"{type(exc).__name__}: {exc}")
        return ProcessResult("failed", base, str(exc),
                             note_path if note_path.exists() else None)

File: apple/ScribedCore/Sources/ScribedCore/Pipeline.swift
import Foundation

public enum ProcessStatus: String, Equatable {
    case done
    case skipped
    case deferredNeedLocal = "deferred_need_local"
    case failed
}

public struct ProcessResult: Equatable {
    public let status: ProcessStatus
    public let base: String
    public let message: String
    public var notePath: URL?
    public var transcriptPath: URL?

    public init(status: ProcessStatus, base: String, message: String,
                notePath: URL? = nil, transcriptPath: URL? = nil) {
        self.status = status; self.base = base; self.message = message
        self.notePath = notePath; self.transcriptPath = transcriptPath
    }
}

/// Injectable effects, mirroring the Python `deps` SimpleNamespace seam so the
/// pipeline can be tested without servers, ffmpeg, or AVFoundation.
public struct PipelineDeps {
    public var convertToWav: (URL, URL) async throws -> Void
    public var transcribe: (URL, TranscribeConfig) async throws -> [String: Any]
    public var ollamaReachable: (String) async -> Bool
    public var summarise: (_ transcript: String, _ url: String, _ model: String,
                           _ options: SummariseOptions, _ noteOwner: String,
                           _ userSpeaker: String) async throws -> String

    public init(
        convertToWav: @escaping (URL, URL) async throws -> Void,
        transcribe: @escaping (URL, TranscribeConfig) async throws -> [String: Any],
        ollamaReachable: @escaping (String) async -> Bool,
        summarise: @escaping (String, String, String, SummariseOptions, String, String) async throws -> String
    ) {
        self.convertToWav = convertToWav
        self.transcribe = transcribe
        self.ollamaReachable = ollamaReachable
        self.summarise = summarise
    }

    /// Real dependencies wired to AVFoundation + the HTTP clients.
    public static func live() -> PipelineDeps {
        let whisper = WhisperXClient()
        let ollama = OllamaClient()
        return PipelineDeps(
            convertToWav: { try await AudioConverter.convertToWav(source: $0, dest: $1) },
            transcribe: { try await whisper.transcribe(wavURL: $0, config: $1) },
            ollamaReachable: { await ollama.reachable($0) },
            summarise: { transcript, url, model, options, owner, speaker in
                let prompt = Prompt.build(transcript: transcript, noteOwner: owner, userSpeaker: speaker)
                return try await ollama.generate(url: url, model: model, prompt: prompt, options: options)
            })
    }
}

/// Port of `meeting_pipeline/pipeline.py`.
public enum Pipeline {

    /// Choose the Ollama target (server vs local) or signal deferral.
    static func chooseOllama(
        _ config: Config, reachable: (String) async -> Bool
    ) async -> (url: String, model: String)? {
        let s = config.summarise
        if s.backend == "local" { return (s.local.url, s.local.model) }
        if await reachable(s.server.url) { return (s.server.url, s.server.model) }
        if s.allowLocalFallback { return (s.local.url, s.local.model) }
        return nil
    }

    public static func processOne(
        path: URL, config: Config, deps: PipelineDeps,
        stableChecks: Int = 3, stableDelay: Double = 2.0
    ) async -> ProcessResult {
        let recordingsDir = Config.resolvePath(config.recordingsDir)
        let base = ScribedState.baseFor(recordingsDir: recordingsDir, path: path)
        let notesDir = Config.resolvePath(config.notesDir)
        let workDir = Config.resolvePath(config.workDir)

        let state: ScribedState.Store
        do {
            state = try ScribedState.Store(
                stateDir: workDir.appendingPathComponent(".state"), notesDir: notesDir)
        } catch {
            return ProcessResult(status: .failed, base: base,
                                 message: "state init failed: \(error.localizedDescription)")
        }

        if state.isDone(base) {
            return ProcessResult(status: .skipped, base: base,
                                 message: "already processed", notePath: state.notePath(base))
        }
        if !ScribedState.waitUntilStable(path, checks: stableChecks, delay: stableDelay) {
            return ProcessResult(status: .skipped, base: base, message: "file still changing")
        }
        if state.isProcessing(base) {
            return ProcessResult(status: .skipped, base: base, message: "already being processed")
        }

        guard let target = await chooseOllama(config, reachable: deps.ollamaReachable) else {
            return ProcessResult(status: .deferredNeedLocal, base: base,
                                 message: "Server Ollama offline; local fallback not allowed yet")
        }

        state.markProcessing(base)
        try? FileManager.default.createDirectory(at: notesDir, withIntermediateDirectories: true)
        let notePath = state.notePath(base)

        do {
            let wavPath = workDir.appendingPathComponent("\(base).wav")
            try await deps.convertToWav(path, wavPath)

            let result = try await deps.transcribe(wavPath, config.transcribe)
            let clean = TranscriptCleaner.clean(TranscriptCleaner.segments(from: result))
            if clean.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                state.markFailed(base, "empty transcript")
                return ProcessResult(status: .failed, base: base, message: "empty transcript")
            }
            let transcriptPath = workDir.appendingPathComponent("\(base).transcript.clean.txt")
            try? (clean + "\n").write(to: transcriptPath, atomically: true, encoding: .utf8)

            let summary = try await deps.summarise(
                clean, target.url, target.model, config.summarise.options,
                config.noteOwner, config.userSpeaker)
            try summary.write(to: notePath, atomically: true, encoding: .utf8)

            let failures = SummaryValidator.validate(summary)
            if !failures.isEmpty {
                let message = failures.joined(separator: "; ")
                state.markFailed(base, message)
                return ProcessResult(status: .failed, base: base, message: message,
                                     notePath: notePath, transcriptPath: transcriptPath)
            }
            state.markDone(base)
            return ProcessResult(status: .done, base: base, message: "note written",
                                 notePath: notePath, transcriptPath: transcriptPath)
        } catch {
            state.markFailed(base, "\(error)")
            let np = FileManager.default.fileExists(atPath: notePath.path) ? notePath : nil
            return ProcessResult(status: .failed, base: base, message: "\(error)", notePath: np)
        }
    }
}

/// Process all pending recordings once (mirrors `WatcherController.scan_once`).
public enum Scanner {
    public static func scanOnce(
        config: Config, deps: PipelineDeps,
        stableChecks: Int = 3, stableDelay: Double = 2.0
    ) async -> [ProcessResult] {
        let recordingsDir = Config.resolvePath(config.recordingsDir)
        let workDir = Config.resolvePath(config.workDir)
        let notesDir = Config.resolvePath(config.notesDir)
        guard let state = try? ScribedState.Store(
            stateDir: workDir.appendingPathComponent(".state"), notesDir: notesDir) else { return [] }
        let pending = ScribedState.iterPending(recordingsDir: recordingsDir, state: state)
        var results: [ProcessResult] = []
        for path in pending {
            results.append(await Pipeline.processOne(
                path: path, config: config, deps: deps,
                stableChecks: stableChecks, stableDelay: stableDelay))
        }
        return results
    }
}
